### PR TITLE
[aievec] Add AIEml intrinsic for matmul ops

### PR DESCRIPTION
**Pull request by jsetoain**

### CODE REVIEW

**Review by david-vc — CHANGES_REQUESTED** [review]
@jsetoain : I request to generalize the new Operation so that it can model all matmul configs that the architecture supports. Thanks.

File: test/dialect/AIEVec/roundtrip.mlir [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+// CHECK-LABEL: @matmul_i32
+// CHECK-SAME: %[[A:.*]]: vector<4x2xi32>
+// CHECK-SAME: %[[B:.*]]: vector<2x4xi32>
+// CHECK-SAME: %[[C:.*]]: vector<4x4xi64>
+// CHECK:      %[[RES:.*]] = aievec.matmul %[[A]], %[[B]], %[[C]] :
+// CHECK-SAME: vector<4x2xi32>, vector<2x4xi32> into vector<4x4xi64>
+// CHECK: return %[[RES]] : vector<4x4xi64>
+func.func @matmul_i32(%A : vector<4x2xi32>, %B : vector<2x4xi32>,
+                      %C : vector<4x4xi64>) -> vector<4x4xi64> {
+  %0 = aievec.matmul %A, %B, %C : vector<4x2xi32>, vector<2x4xi32>
```
Comment (david-vc) [inline test/dialect/AIEVec/roundtrip.mlir:92]: This combination does not appear in the table embedded in AIEVec_MatMulOp (AIEVecOps.td). The only combination with 4x2xi32 for A, uses 2x4xi16 for B and using a 4x4xi64 acc.
So is this test really passing, or is there a problem with the validation code?

File: test/dialect/AIEVec/roundtrip.mlir [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+// CHECK-LABEL: @matmul_f32
+// CHECK-SAME: %[[A:.*]]: vector<4x8xf32>
+// CHECK-SAME: %[[B:.*]]: vector<8x4xf32>
+// CHECK-SAME: %[[C:.*]]: vector<4x4xf32>
+// CHECK:      %[[RES:.*]] = aievec.matmul %[[A]], %[[B]], %[[C]] :
+// CHECK-SAME: vector<4x8xf32>, vector<8x4xf32> into vector<4x4xf32>
+// CHECK: return %[[RES]] : vector<4x4xf32>
+func.func @matmul_f32(%A : vector<4x8xf32>, %B : vector<8x4xf32>,
+                      %C : vector<4x4xf32>) -> vector<4x4xf32> {
+  %0 = aievec.matmul %A, %B, %C : vector<4x8xf32>, vector<8x4xf32>
```
Comment (david-vc) [inline test/dialect/AIEVec/roundtrip.mlir:28]: This combination does not appear in the table embedded in AIEVec_MatMulOp (AIEVecOps.td). There are not combinations with f32 operand listed.
So is this test really passing, or is there a problem with the validation code?

**Review by david-vc — COMMENTED** [review]
This is the first AIEVec op that requires its operands to be 2-dimensional. I think this direction is worth pursuing.

File: include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+//
+
+//===----------------------------------------------------------------------===//
+// Extra type constraint definitions for AIEVec operations.
+//===----------------------------------------------------------------------===//
+
+include "mlir/IR/BuiltinTypes.td"
+include "mlir/IR/OpBase.td"
+
+def I4  : I<4>;
```
Comment (jamestcl-amd) [inline include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td:17]: Just curious, so only I4 needs to be defined at here?

File: include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                       string op2, Type t2,
+                       string op3, Type t3> :
+  And<[SubstLeaves<"$_self", VectorType<op1>.result, t1.predicate>,
+       SubstLeaves<"$_self", VectorType<op2>.result, t2.predicate>,
+       SubstLeaves<"$_self", VectorType<op3>.result, t3.predicate>]>;
+
+class IsValidAIE2MatMulShapeAndType<string lhs, string rhs, string acc> :
+  PredOpTrait<lhs # " x " # rhs # " = " # acc # " is a valid AIE2 " #
+              "matrix-multiply and accumulate op",
+              Or<[VectorTypesMatch<lhs, AIE2MatMulOperand_4x16xi8,
```
Comment (jamestcl-amd) [inline include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td:129]: These checks and constraints look so nice!

File: test/dialect/AIEVec/roundtrip.mlir (hunk 1 above)
Comment (jsetoain) [reply test/dialect/AIEVec/roundtrip.mlir:92]: Yes, tests need updating.

**Review by david-vc — APPROVED** [review]

File: include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+//         be a ShapedType
+def AIE2MatMulOperand_4x16xi8  : VectorOfShapeAndType<[4, 16], I8>;
+def AIE2MatMulOperand_4x8xi8   : VectorOfShapeAndType<[4, 8], I8>;
+def AIE2MatMulOperand_4x4xi16  : VectorOfShapeAndType<[4, 4], I16>;
+def AIE2MatMulOperand_4x2xi16  : VectorOfShapeAndType<[4, 2], I16>;
+def AIE2MatMulOperand_2x8xi16  : VectorOfShapeAndType<[2, 8], I16>;
+def AIE2MatMulOperand_4x8xi16  : VectorOfShapeAndType<[4, 8], I16>;
+def AIE2MatMulOperand_2x4xi16  : VectorOfShapeAndType<[2, 4], I16>;
+def AIE2MatMulOperand_4x2xi32  : VectorOfShapeAndType<[4, 2], I32>;
+def AIE2MatMulOperand_4x8xbf16 : VectorOfShapeAndType<[4, 8], BF16>;
```
Comment (makslevental) [inline include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td:52]: Nit: is there a reason not to just "forward" the `VectorOfShapeAndType<...>` to the use site? I.e., I'm asking can't you use `VectorOfShapeAndType<[4, 8], BF16>` in the type constraint below?

File: include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+class VectorOfShapeAndType<list<int> shape, Type type> :
+  Type<And<[VectorOfShape<shape>,
+            VectorOfElementType<type>]>,
+        "vector of shape <" # !interleave(shape, "x") # "> and",
+        "::mlir::VectorType">;
+
+
+// Notice: These predicate definitions assume that the type has been verified to
+//         be a ShapedType
+def AIE2MatMulOperand_4x16xi8  : VectorOfShapeAndType<[4, 16], I8>;
```
[the remark this answers is not in the document]
Comment (makslevental) [reply include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td:44]: Yea I agree, something like `SmallMatrix...` maybe.

File: include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+def AIE2MatMulOperand_2x8xi16  : VectorOfShapeAndType<[2, 8], I16>;
+def AIE2MatMulOperand_4x8xi16  : VectorOfShapeAndType<[4, 8], I16>;
+def AIE2MatMulOperand_2x4xi16  : VectorOfShapeAndType<[2, 4], I16>;
+def AIE2MatMulOperand_4x2xi32  : VectorOfShapeAndType<[4, 2], I32>;
+def AIE2MatMulOperand_4x8xbf16 : VectorOfShapeAndType<[4, 8], BF16>;
+
+def AIE2MatMulOperand_16x8xi4  : VectorOfShapeAndType<[16, 8], I4>;
+def AIE2MatMulOperand_8x8xi8   : VectorOfShapeAndType<[8, 8], I8>;
+def AIE2MatMulOperand_8x4xi8   : VectorOfShapeAndType<[8, 4], I8>;
+def AIE2MatMulOperand_8x4xbf16 : VectorOfShapeAndType<[8, 4], BF16>;
```
Comment (makslevental) [inline include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td:57]: Is there a reason not to just "forward" the `VectorOfShapeAndType<...>` to the use site? I.e., I'm asking can't you use `VectorOfShapeAndType<[4, 8], BF16>` in the type constraint below?

File: include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+             AIE2MatMulOperand_2x8xi64,
+             AIE2MatMulOperand_4x4xi64,
+             AIE2MatMulOperand_4x4xf32],
+            "a vector compatible with an accumulator of matrix-multiply and "
+            # "accumulate",
+            "::mlir::VectorType">;
+
+class ShapeDimsMatch<string lhs, int ld, string rhs, int rd> :
+  CPred<Shape<lhs>.result # "[" # ld # "] == " #
+        Shape<rhs>.result # "[" # rd # "]">;
```
Comment (makslevental) [inline include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td:105]: I thought this existed upstream but I can't find it - good stuff for upstreaming!

File: include/aie/Dialect/AIEVec/IR/CMakeLists.txt [hunk 9]
```diff
@@ -6,4 +6,4 @@
 # (c) Copyright 2022 Xilinx Inc.
 
 add_mlir_dialect(AIEVecOps aievec)
-add_mlir_doc(AIEVecOps AIEVecDialect ./ -gen-dialect-doc)
+add_mlir_doc(AIEVecOps AIEVecDialect ./ -gen-dialect-doc -dialect=aievec)
```
Comment (makslevental) [inline include/aie/Dialect/AIEVec/IR/CMakeLists.txt:9]: Gonna have a collision when you rebase :)

**Review by makslevental — COMMENTED** [review]
LGTM (just some curious questions).

File: include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td (hunk 7 above)
Comment (jsetoain) [reply include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td:57]: Progressive refinement. I used to have them separated by lhs/rhs/acc even, but it's true that at this point I may as well remove the "alias".

File: include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td (hunk 8 above)
Comment (jsetoain) [reply include/aie/Dialect/AIEVec/IR/AIEVecTypeConstraints.td:105]: Not this one, but you just reminded me that "VectorOfShape" does exist in newer versions of MLIR (we were stuck in an old one for quite some time), I should get rid of that one.

File: include/aie/Dialect/AIEVec/IR/CMakeLists.txt (hunk 9 above)
Comment (jsetoain) [reply include/aie/Dialect/AIEVec/IR/CMakeLists.txt:9]: I know, but it would pass CI otherwise :-)

**Pull request merged**